File: js/task3/S3_PIPE_06.js
// S3_PIPE_06 — Safe pipeline

function pipeSafe(...fns) {
  return (input) => {
    let value = input;

    try {
      for (const fn of fns) {
        value = fn(value);
      }
      return { ok: true, value };
    } catch (error) {
      return { ok: false, error: error instanceof Error ? error.message : String(error) };
    }
  };
}

// Tests
const parseNumber = (x) => {
  const n = Number(x);
  if (Number.isNaN(n)) throw new Error("Not a number");
  return n;
};

const inverse = (n) => {
  if (n === 0) throw new Error("Division by zero");
  return 1 / n;
};

const safeProcess = pipeSafe(parseNumber, inverse);

console.log(safeProcess("4"));   // { ok: true, value: 0.25 }
console.log(safeProcess("0"));   // { ok: false, error: 'Division by zero' }
console.log(safeProcess("x"));   // { ok: false, error: 'Not a number' }

File: python/task3c/S3_PIPE_06.py
# S3_PIPE_06 — Safe pipeline (pipe_safe)

def pipe_safe(*fns):
    def runner(x):
        value = x
        try:
            for fn in fns:
                value = fn(value)
            return {"ok": True, "value": value}
        except Exception as e:
            return {"ok": False, "error": str(e)}
    return runner


def main():
    def to_int(s):
        return int(s)  # raises ValueError on bad input

    def double(n):
        return n * 2

    safe = pipe_safe(to_int, double)

    print("good:", safe("21"))   # ok True, value 42
    print("bad:", safe("xx"))    # ok False, error message


if __name__ == "__main__":
    main()
